libs/net/nodenetworkconfigurationpolicy.py

```diff
@@ -0,0 +1,140 @@
+
+
+@dataclass
+class Interface:
+    name: str
+    type: str
+    state: str
+    ipv4: IPv4 | None = None
+    ipv6: IPv6 | None = None
+    bridge: Bridge | None = None
+
+
+@dataclass
+class DesiredState:
+    """
+    Represents the desired network configuration for NMstate.
+    Following the NMstate YAML-based API specification:
+    https://nmstate.io/devel/yaml_api.html
+    """
+
+    interfaces: list[Interface]
+
+
+class NodeNetworkConfigurationPolicy(Nncp):
+    """
+    NodeNetworkConfigurationPolicy object.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        desired_state: DesiredState,
+        node_selector: dict[str, str] | None = None,
+    ):
+        """
+        Create and manage NodeNetworkConfigurationPolicy
+
+        Args:
+            name (str): Name of the NodeNetworkConfigurationPolicy object.
+            desired_state (DesiredState): Desired policy configuration - interface creation, modification or removal.
+            node_selector (dict, optional): A node selector that specifies the nodes to apply the node network
+                configuration policy to.
+        """
+        self._desired_state = desired_state
+        super().__init__(
+            name=name,
+            desired_state=asdict(desired_state, dict_factory=self._dict_normalization),
+            node_selector=node_selector,
+        )
+
+    @staticmethod
+    def _dict_normalization(data: list[tuple[str, Any]]) -> dict[str, Any]:
+        """Filter out none values and converts key characters containing underscores into dashes."""
+        return {key.replace("_", "-"): val for (key, val) in data if val is not None}
+
+    @property
+    def desired_state_spec(self) -> DesiredState:
+        return self._desired_state
+
+    def clean_up(self) -> bool:
+        self._delete_interfaces()
+        self.wait_for_status_success()
+        return super().clean_up()
+
+    @retry(
+        wait_timeout=WAIT_FOR_STATUS_TIMEOUT_SEC,
+        sleep=WAIT_FOR_STATUS_INTERVAL_SEC,
+        exceptions_dict={AssertionError: []},  # Using a no-op exception so any other error will be raised
+    )
+    def wait_for_status_success(self) -> dict[str, Any]:
+        conditions = (
+            condition
+            for condition in self.instance.status.conditions
+            if condition["status"] == Nncp.Condition.Status.TRUE
+        )
+
+        for condition in conditions:
+            if condition["type"] == Nncp.Condition.AVAILABLE:
+                self.logger.info(f"{self.kind}/{self.name} configured successfully")
+                return condition
+            if condition["type"] == Nncp.Condition.DEGRADED:
+                raise NNCPConfigurationFailed(f"{self.name} failed on condition:\n{condition}")
+        return {}
+
+    def _delete_interfaces(self) -> None:
+        desired_state = deepcopy(self.desired_state)
+        for iface in desired_state["interfaces"]:
+            iface["state"] = Nncp.Interface.State.ABSENT
```

Comment:
Maybe we should limit that only for specific interface types.
If the original NNCP is supposed to modify something in a node's interface, e.g. setting `eno3` to `down`, we wouldn't want this scenario to end with an attempt to delete `eno3`.
And more generally - we don't want to try deleting any interface that has always been there.

Reply:
You are right, but this is following the current logic and going beyond that is a non-trivial follow-up.

In general, what you suggest here is possible.
Do we need it? Only time will tell (or you will tell us :) ).

Reply:
After discussing it offline with @EdDev - I can live with this change as it is presented above for 2 reasons:
- It actually preserves the same logic we used until now, which haven't resulted with any attempt to delete any day-0 interface.
- Checking the interface type might not be enough, for example - the default `br-ex` is of type `ovs-bridge`, and we shouldn't attempt to remove it.

Reply:
Marking this conversation as resolved.